src/dimension_api/__init__.py

```diff
@@ -0,0 +1,15 @@
+"""Dimension API."""
+
+from typing import Final, Protocol, Self, runtime_checkable
+
+__version__: Final = "0.0.1.dev0"
+__all__ = ["__version__", "Dimension"]
+
+@runtime_checkable
+class Dimension(Protocol):
+    @property
+    def is_dimensionless(self) -> bool: ...
```

Comment:
I think we can make this simpler: I agree a `@property` should satisfy the protocol, but we can also accept a static attributes, right ?
```suggestion
    is_dimensionless: bool
```

Reply:
Reading https://docs.python.org/3/tutorial/classes.html#class-and-instance-variables, class variables (which is what these static attributes are, right?) are shared by all instances. But some instances should be dimensionless, while others aren't, no?

Reply:
Declaring instance attributes at the class level is a very common idiom (that's how stdlib's `@dataclass` and `class MyTuple(NamedTuple)` are *supposed* to be used) and unfortunately there is no good way to annotate that we *do* mean it to be an instance attribute, other than *not* annotating as [`ClassVar`](https://docs.python.org/3/library/typing.html#typing.ClassVar`). I'll run a couple simple tests and come back to you in a minute.

Reply:
Also, static attributes mean the protocol cannot be used for `issubclass` checks!

> isinstance() can be used with both data and non-data protocols, while issubclass() can be used only with non-data protocols. This restriction exists because some data attributes can be set on an instance in constructor and this information is not always available on the class object.
https://peps.python.org/pep-0544/

Reply:
```python
# t.py
from typing import Protocol, ClassVar

class Person:
    name: str

class Group:
    name: ClassVar[str]

class Animal:
    @property
    def name(self) -> str: ...

# first test
class HasName1(Protocol):
    name: str

def foo(named: HasName1) -> None:
    ...

foo(Person())
foo(Group())
foo(Animal())

# second test
class HasName2(Protocol):
    @property
    def name(self) -> str: ...

def bar(named: HasName2) -> None:
    ...

bar(Person())
bar(Group())
bar(Animal())
```

```
❯ uvx pyright t.py
/private/tmp/classvar/t.py
  /private/tmp/classvar/t.py:21:5 - error: Argument of type "Group" cannot be assigned to parameter "named" of type "HasName1" in function "foo"
    "Group" is incompatible with protocol "HasName1"
      "name" is not defined as a ClassVar in protocol (reportArgumentType)
  /private/tmp/classvar/t.py:22:5 - error: Argument of type "Animal" cannot be assigned to parameter "named" of type "HasName1" in function "foo"
    "Animal" is incompatible with protocol "HasName1"
      "name" is invariant because it is mutable
      "name" is an incompatible type
        "property" is not assignable to "str" (reportArgumentType)
  /private/tmp/classvar/t.py:33:5 - error: Argument of type "Group" cannot be assigned to parameter "named" of type "HasName2" in function "bar"
    "Group" is incompatible with protocol "HasName2"
      "name" is not defined as a ClassVar in protocol (reportArgumentType)
3 errors, 0 warnings, 0 informations
```

conclusion: `@property` confers an additional meaning of immutability, which can't be expressed with a simple instance attribute, and it's really what we want here, so please ignore my initial suggestion !

Reply:
I think this attribute name is worth putting up for a vote as pint uses `.dimensionless`

Reply:
Since astropy never defined a dimension class, I don't mind about the name. 

But thinking a little more, it seems not illogical to decide that the `dimension` object is truthy only if it is not dimensionless. The meaning would seem fairly obvious, since dimensionless would be like an empty list, etc.

If that is not explicit enough but we're want to punt on having an attribute, one could just test equality, i.e., `given_dimension == (dimensionless := Dimension())`. Indeed, with that, I do wonder why one would need a specific attribute at all.

Reply:
Agreed on the motivation. If there were a more fundamental method/property we could add that'd also be good.